src/dbnode/storage/forward_index_dice.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/x/dice"
+)
+
+// forwardIndexDice is a die roll that adds a chance for incoming index writes
+// arriving near a block boundary to be duplicated and written to the next block
+// index, adding jitter and smoothing index load.
+type forwardIndexDice struct {
+	enabled   bool
+	blockSize time.Duration
+
+	forwardIndexThreshold time.Duration
+	forwardIndexDice      dice.Dice
+}
+
+func newForwardIndexDice(
+	opts Options,
+) (forwardIndexDice, error) {
+	var (
+		indexOpts  = opts.IndexOptions()
+		seriesOpts = opts.SeriesOptions()
+
+		probability = indexOpts.ForwardIndexProbability()
+	)
+
+	// NB: if not enabled, return a no-op forward index dice.
+	if probability == 0 {
+		return forwardIndexDice{}, nil
+	}
+
+	var (
+		threshold = indexOpts.ForwardIndexThreshold()
+
+		retention    = seriesOpts.RetentionOptions()
+		bufferFuture = retention.BufferFuture()
+		blockSize    = retention.BlockSize()
+
+		forwardIndexThreshold time.Duration
+	)
+
+	if threshold > 1 {
+		return forwardIndexDice{},
+			fmt.Errorf("invalid forward write threshold %f", threshold)
+	}
+
+	bufferFragment := float64(bufferFuture) * threshold
```

Comment:
I don't understand why there is a relationship between bufferFuture and the threshold...I guess in my mind bufferFuture would set a maximum on the threshold so if bufferFuture is 10 minutes and you have a block size of 2 hours then the earliest you could start forward indexing would be at 1:50PM, but I don't understand the logic here where you multiply bufferFuture by some threshold

Reply:
Yeah, so threshold is defined as a fragment of the bufferFuture to allow some tweaking on when ; my thinking is you'd put the `threshold` at 0.9 or something, and then in your example, we'd start forward indexing at `1:51 PM`, just to cut down on any possibility of trying to write the values too early in case of positive skew on the emitter.

i.e. `bufferFragment ∈ (0, bufferFuture]`

Reply:
mind adding an explanation for this in a comment